rfcs/0171-fetch-from-github.md

```diff
@@ -0,0 +1,114 @@
+  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
+                                 Dload  Upload   Total   Spent    Left  Speed
+100   173  100   173    0     0    754      0 --:--:-- --:--:-- --:--:--   755
+100  627k    0  627k    0     0   604k      0 --:--:--  0:00:01 --:--:-- 1014k
+unpacking source archive /build/8c542627d936a01b1d97825e7f26a8e95633f7aa.tar.gz
+/nix/store/dkvcfm90ckrlgmv89s8sr15vcidwlxhs-openAV-ArtyFX-8c542627d9-src
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- All derivations which don't pass a "name" parameter will need to be re-realized
+    - This will be a download-intensive one-time cost to realize the new FOD derivations.
+    - NAR hash should not need to be recomputed assuming it was deterministic and not stale.
+    - Cache should be minimally impacted as NARs are content addressable, thus deterministic sources should not contribute to cache bloat.
+    - Potential for sources which are no longer available to be broken.
+        - These can have their name manually set to "source" to perserve previous behavior.
+        - Ideally source availability would be remedied with more appropriate methods. E.g. being made available.
+- "Interchangeability" with other fetchers is diminished as the derivation name is different
+    - In practice, fetchFromGitHub is never used in this way. It is generally the only fetcher, so there is never another FOD to dedupilicate.
+- Out-of-tree repositories may get hash mismatch errors
+    - If the cause of the mismatch is staleness, this is good and working as intended
+    - If the cause is non-determinism, this is frustrating.
+- Some derivations assume "source" to be the name of sourceRoot
+    - This has been mitigated over two years within Nixpkgs
+    - Out-of-tree code may break if they assume "source" is the name
+        - Can be mitigated with release notes describing the issue
+
+# Alternatives
+[alternatives]: #alternatives
+
+- Do nothing. Retain current status quo.
+
+- In https://github.com/NixOS/nixpkgs/pull/153386#issuecomment-1007729116, @Ericson2314 mentioned that this may be solved at the Nix tooling level. And that a year should be give to see if an implementation can be done.
+    - That was 2+ years ago, and an ounce of prevention today is worth ten ounces of remedy tomorrow.
+
+# Prior art
+[prior-art]: #prior-art
+
+- https://github.com/NixOS/nixpkgs/pull/153386
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+- Full commit hashes could be truncated. This sacrifices a bit of simplicity for better looking derivation names:
+```
+let
+  version = builtins.replaceStrings [ "refs/tags/" ] [ "" ] rev;
+  # Git commit hashes are 40 characters long, assume that very long versions are version-control labels
+  ref = if (builtins.stringLength rev) > 15 then builtins.substring 0 8 version else version;
+in lib.sanitizeDerivationName "${repo}-${ref}-src";
```

Comment:
I wouldn't recommend this. Short hashes are not guaranteed to be stable for long term storage.

Reply:
The current "source" isn't that stable either. 8 characters is still a fair amount of entropy, and likely to be different enough for most repositories.

Reply:
Also remember, that it doesn't have to "unique across all time". It just needs to be different than what was there before, so that the combination of name + hash are different.